### PR TITLE
spring clean

### PR DESCRIPTION
**Pull request by davidchambers**
It's been a while since anyone gave this project love and attention, so here's some :heart:.

Changes:

  - Replace Ramda with Sanctuary
  - Replace data.task with Fluture
  - Define common internal functions: `exit0`, `exit1`, `join`, and `readFile`
  - Remove `--harmony` flags
  - Remove Babel, which is no longer required now that Node supports [`await`][1]
  - Upgrade dependencies
  - Integrate ESLint and sanctuary-style

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await

### CODE REVIEW

File: bluebird-promisell.js [hunk 1]
```diff
@@ -1,56 +1,26 @@
...
-          process.exit(0);
-        },
-        err => {
-          process.stderr.write(String(err) + '\n');
-          process.exit(1);
-        });
+  concatFiles(join(process.argv[2])).then(exit0, exit1);
 };
 
 if (process.mainModule.filename === __filename) main();
```
Comment (davidchambers) [inline bluebird-promisell.js]: You may be interested in these changes, @zhangchiqing.

File: futures.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+          S.chain(S.traverse(Future, readFile)),  // :: Future Error (Array String)
+          S.map(S.joinWith(''))],                 // :: Future Error String
+         'index.txt');
+
+
+const main = () => {
+  concatFiles(join(process.argv[2])).fork(exit1, exit0);
+};
+
+if (process.mainModule.filename === __filename) main();
```
Comment (davidchambers) [inline futures.js]: Does this file look good to you, @Avaq?

File: kgo.js [hunk 3]
```diff
@@ -1,41 +1,27 @@
...
-  });
+  ('index', readFile(path('index.txt')))
+  ('filePaths', ['index'], kgo.sync(S.compose(S.map(path), S.lines)))
+  ('files', ['filePaths'], S.curry3(foreign.parallel, readFile))
+  ('concated', ['files'], kgo.sync(S.joinWith('')))
+  (['concated'], exit0)
+  (['*'], exit1);
 };
 
 if (process.mainModule.filename === __filename) main();
```
Comment (davidchambers) [inline kgo.js]: You may be interested in these changes, @KoryNunn.

File: lazy-either.js [hunk 4]
```diff
@@ -1,46 +1,52 @@
...
+const exit0         = require('./common/exit0');
+const exit1         = require('./common/exit1');
+const join          = require('./common/join');
+
+
+//    Array$traverse :: Applicative f => TypeRep f -> (a -> f b) -> Array a -> f (Array b)
+//
+//    We cannot use S.traverse because LazyEither does not currently support
+//    recent versions of the Fantasy Land specification. It's thus necessary
+//    to dispatch to `map` and `ap` (rather than their prefixed equivalents).
```
Comment (davidchambers) [inline lazy-either.js]: Are you still maintaining LazyEither, @Risto-Stevcev? If so, have you plans to support `fantasy-land/`-prefixed methods?

File: most.js [hunk 5]
```diff
@@ -1,42 +1,28 @@
...
-    process.stdout.write(value);
-    process.exit(0);
-  }, err => {
-    process.stderr.write(String(err) + '\n');
-    process.exit(1);
-  });
+  concatFiles(join(process.argv[2])).then(exit0, exit1);
 };
 
 if (process.mainModule.filename === __filename) main();
```
Comment (davidchambers) [inline most.js]: You may be interested in these changes, @Risto-Stevcev.

File: promises-pipe.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+          then(Promise.all.bind(Promise)),
+          then(S.joinWith(''))],
+         'index.txt');
+
+
+const main = () => {
+  concatFiles(join(process.argv[2])).then(exit0, exit1);
+};
+
+if (process.mainModule.filename === __filename) main();
```
Comment (davidchambers) [inline promises-pipe.js]: You may be interested in these changes, @paldepind.

File: righto.js [hunk 7]
```diff
@@ -1,31 +1,30 @@
...
-      process.stderr.write(String(err) + '\n');
-      process.exit(1);
-    } else {
-      process.stdout.write(data);
-      process.exit(0);
-    }
+  concatFiles(join(process.argv[2]))((err, data) => {
+    if (err == null) exit0(data); else exit1(err);
   });
 };
```
Comment (davidchambers) [inline righto.js]: You may be interested in these changes, @KoryNunn.

File: futures.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+const readFile =
+S.compose(Future.node, S.curry3(fs.readFile, S.__, {encoding: 'utf8'}));
+
+//    concatFiles :: (String -> String) -> Future Error String
+const concatFiles = path =>
+  S.pipe([path,                                   // :: String
+          readFile,                               // :: Future Error String
+          S.map(S.lines),                         // :: Future Error (Array String)
+          S.map(S.map(path)),                     // :: Future Error (Array String)
+          S.chain(S.traverse(Future, readFile)),  // :: Future Error (Array String)
```
Comment (Avaq) [inline futures.js]: If it's important to read all files in parallel, we can replace this line with one of two versions:

1. Using `traverse` over ConcurrentFuture (Par) structure:

```js
S.map(S.map(Future.Par)),
S.chain(S.compose(Future.seq, S.traverse(Future.Par, S.compose(Future.Par,  readFile))))
```

2. Using `Future.parallel`:

```js
S.map(S.map(readFile)),
S.chain(Future.parallel(Infinity))
```

Comment (davidchambers) [inline futures.js]: Reading files in parallel would be very nice indeed! Which version do you favour, and why? I'm happy to follow your recommendation.

Comment (Avaq) [inline futures.js]: It saddens me\* to prefer the second version, because it's more concise and it's clearer what's going on. It's also more similar to many of the Promise based solutions in the project.

\* It would be nicer to be able to utilize `traverse` for this purpose without the verbose type juggling. If we would go for the first approach, some ways to make it more readable would be to name `S.compose(Future.Par,  readFile)` something like `concurrentReadFile`, and maybe use `map` to avoid the nested `compose`:

```js
const concurrentReadFile = S.compose(Future.Par, readFile);
//...
  S.map(S.traverse(Future.Par, concurrentReadFile)), // :: Future Error (ConcurrentFuture Error (Array String))
  S.chain(Future.seq)                                // :: Future Error (Array String)
```

File: common/sanctuary.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+
+//    ConcurrentFutureType :: Type -> Type -> Type
+const ConcurrentFutureType = $.BinaryType(
+  Future.Par['@@type'],
+  'https://github.com/fluture-js/Fluture#concurrentfuture',
+  x => S.type(x) === Future.Par['@@type'],
+  f => [],
+  f => []
+);
```
Comment (davidchambers) [inline common/sanctuary.js]: Does this look right to you, @Avaq? It seems the ConcurrentFuture type lacks helpers corresponding to those that exist for defining `FutureType :: Type -> Type -> Type`. I'd love to be able to do this:

```javascript
const Future = require('fluture');
const $ = require('sanctuary-def');

const {ConcurrentFutureType, FutureType} = Future.defineTypes($);
```

Fluture could then provide type definitions *without* depending on sanctuary-def (although it would be a good idea to specify sanctuary-def as a peer dependency). What do you think?

File: kgo.js (hunk 3 above)
Comment (KoryNunn) [inline kgo.js]: Honestly I'm not advocating for kgo's usage these days, but rather suggest people look at righto. I'd be happy to drop this case.

File: righto.js (hunk 7 above)
Comment (KoryNunn) [inline righto.js]: The change to remove variables for each logical step goes violently against the grain of righto, and makes this, to my eyes, unreadable.

Also the change to use sanctuary within this example in my opinion muddies it, since a new viewer would have to look up how it works AND how righto works.

At this point, since all the examples use sanctuary, I don't see this as a useful comparison, but a contrived example that is designed explicitly to show of one or two particular examples.

Comment (davidchambers) [inline righto.js]: I updated the implementation. Is this more to your liking?

```javascript
//    concatFiles :: (String -> String) -> Righto String
const concatFiles = path => {
  const readFileRel = S.compose(readFile, path);
  const index = readFileRel('index.txt');
  const files = righto.sync(S.compose(S.map(readFileRel), S.lines), index);
  return righto.sync(S.joinWith(''), righto.all(files));
};
```

File: kgo.js (hunk 3 above)
Comment (davidchambers) [inline kgo.js]: Okay. Done. :)

File: righto.js (hunk 7 above)
Comment (KoryNunn) [inline righto.js]: That's definitely much more like how it is intended to be used.

Comment (davidchambers) [inline righto.js]: Okay. Thank you for your feedback, Kory. :)

File: common/sanctuary.js [hunk 10]
```diff
@@ -5,10 +5,19 @@ const S             = require('sanctuary');
 const $             = require('sanctuary-def');
 
 
+//    ConcurrentFutureType :: Type -> Type -> Type
+const ConcurrentFutureType = $.BinaryType(
+  Future.Par['@@type'],
+  'https://github.com/fluture-js/Fluture#concurrentfuture',
+  x => S.type(x) === Future.Par['@@type'],
+  f => [],
+  f => []
```
Comment (Avaq) [inline common/sanctuary.js]: I suppose you could use `compose(Future.extractLeft, Future.seq)`.

File: lazy-either.js (hunk 4 above)
Comment (Risto-Stevcev) [inline lazy-either.js]: It's implementing the fantasy land spec, what are the fantasy-land prefixed methods?

It's not actively developed but it's lazily maintained (accepting PRs). Avaq's fluture lib handles the basic cases for using Either and Maybe with futures so I recommend that to people. Due to LazyEither's restrictions, it does have some interesting properties that fluture can't have because it's more general, but I'm more interested in purescript right now.

File: common/sanctuary.js (hunk 9 above)
Comment (Avaq) [inline common/sanctuary.js]: > I'd love to be able to [use something like `Future.defineTypes($)`]

https://github.com/fluture-js/Fluture/issues/101

File: common/sanctuary.js (hunk 10 above)
Comment (davidchambers) [inline common/sanctuary.js]: :zap:

File: lazy-either.js (hunk 4 above)
Comment (davidchambers) [inline lazy-either.js]: > It's implementing the fantasy land spec, what are the fantasy-land prefixed methods?

Methods have required the `fantasy-land/` prefix since fantasyland/fantasy-land#146 which was released in [v1.0.0][]. The current version is [v3.2.0][], so LazyEither is a little out of date. I'll submit a pull request to prefix the methods. :)


[v1.0.0]: https://github.com/fantasyland/fantasy-land/releases/tag/v1.0.0
[v3.2.0]: https://github.com/fantasyland/fantasy-land/releases/tag/v3.2.0

Comment (davidchambers) [inline lazy-either.js]: Risto-Stevcev/lazy-either#1

File: bluebird-promisell.js (hunk 1 above)
Comment (zhangchiqing) [inline bluebird-promisell.js]: Nice refactor! 

For the `concatFiles` function, it's very clean to chain all the sub functions together. However I'm afraid it would be a bit hard for learners to read.

I'd like to add some intermediate variables and type signatures so that it's clear what each step returns. 
And this style also highlights that the logic reads as if it's imperative and sync, but is actually functional and async. 

```js
//  concatFiles :: (String -> String) -> Promise Error String
const concatFiles = path => {
  // String -> Promise Error String
  const readFileFromPath = S.compose(readFile, path);
  // Promise Error String
  const indexFileP = readFileFromPath('index.txt');
  // Promise Error [String]
  const fileNamesP = P.liftp1(S.lines)(indexFileP);
  // Promise Error [String]
  const filesP = P.liftp1(P.traversep(readFileFromPath)(fileNamesP);
  // Promise Error String
  return P.liftp1(S.joinWith(''))(filesP);
}
```

Comment (davidchambers) [inline bluebird-promisell.js]: Thanks for the feedback, @zhangchiqing. I pushed an update. Let me know what you think of it. :)

Comment (zhangchiqing) [inline bluebird-promisell.js]: Looks great! Thanks @davidchambers 👍

File: futures.js (hunk 2 above)
Comment (Avaq) [inline futures.js]: :white_check_mark:

File: futures.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ const S             = require('./common/sanctuary');
 
 //    readFile :: String -> Future Error String
 const readFile =
-S.compose(Future.node, S.curry3(fs.readFile, S.__, {encoding: 'utf8'}));
+S.curry3(Future.encaseN2, fs.readFile, S.__, {encoding: 'utf8'});
```
Comment (Avaq) [inline futures.js]: I prefer the more concise `S.flip(Future.encaseN2(fs.readFile), {encoding: 'utf8'})`, why did you choose this option?

Comment (davidchambers) [inline futures.js]: I find the placeholder clearer than flipping in this case. I'm happy to change it, though.

File: most.js [hunk 12]
```diff
@@ -1,42 +1,28 @@
...
-         R.reduce(R.concat, '')
-  )(join(dir, 'index.txt'));
+//    concatFiles :: (String -> String) -> Promise Error String
+const concatFiles = path =>
+  most.fromPromise(readFile(path('index.txt')))
+  .map(S.lines)
+  .map(S.map(path))
+  .map(S.map(readFile))
+  .map(Promise.all.bind(Promise))
+  .awaitPromises()
```
Comment (Avaq) [inline most.js]: I think it's weird to use `Promise.all` followed by `awaitPromises` to await *a single* Promise in the context of streaming.

I think we could use `most.from` to create a stream of multiple promises, and then await all of them. That also avoids the joining of an array of strings followed by the joining of a single string to an empty string.

```js
//    concatFiles :: String -> Promise Error String
const concatFiles = path =>
  most.fromPromise(readFile(path('index.txt'))) // :: Stream Error String
  .map(S.lines)                                 // :: Stream Error (Array String)
  .map(S.map(path))                             // :: Stream Error (Array String)
  .map(S.map(readFile))                         // :: Stream Error (Array (Promise Error String))
  .chain(most.from)                             // :: Stream Error (Promise Error String)
  .awaitPromises()                              // :: Stream Error String
  .reduce(S.joinWith(''), '')                   // :: Promise Error String
```

File: await.js [hunk 13]
```diff
@@ -1,45 +1,23 @@
...
-    process.stderr.write(String(err) + '\n');
-    process.exit(1);
+    exit1(err);
   }
-  process.stdout.write(results.join(''));
-};
+  Promise.all(S.map(readFileRel, S.lines(index)))
+  .then(S.joinWith(''))
+  .then(exit0, exit1);
+}
```
Comment (davidchambers) [inline await.js]: :zap:

File: most.js (hunk 12 above)
Comment (davidchambers) [inline most.js]: This seems much better (although the reducing function should be `S.concat` rather than `S.joinWith('')`). When I try it in one of the failure cases, though, I see:

  - `UnhandledPromiseRejectionWarning: Unhandled promise rejection`
  - `DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

File: lazy-either.js (hunk 4 above)
Comment (davidchambers) [inline lazy-either.js]: I'll remove the LazyEither solution, Risto. As I stated in Risto-Stevcev/lazy-either#2, Fluture and LazyEither are duplicative. There's little value in including both here.

File: most.js (hunk 12 above)
Comment (Avaq) [inline most.js]: Ah yes - I got confused. I think `S.joinWith('')` is equal to `reduce(S.concat, '')`.

Do you know which Promise is failing? The `index.txt` or one of the others?

Comment (davidchambers) [inline most.js]: > Do you know which Promise is failing?

No. Determining this will require investigation.

File: promises-pipe.js (hunk 6 above)
Comment (paldepind) [inline promises-pipe.js]: Nice 😄 Why did you change `then(Promise.all)` to `then(Promise.all.bind(Promise))`? It should work fine without the bind I think.

Comment (davidchambers) [inline promises-pipe.js]: I switched from Bluebird promises to native promises, which require context binding.

File: most.js (hunk 12 above)
Comment (Avaq) [inline most.js]: The problem is that the Promise settles before Most attaches handlers. In order to make sure handlers are attached, we use `most.fromPromise` immediately on the return value from `readFile`:

```js
//    readFileStream :: String -> Stream Error String
const readFileStream = S.compose(most.fromPromise, readFile);

//    concatFiles :: (String -> String) -> Promise Error String
const concatFiles = path =>
  most.fromPromise(readFile(path('index.txt'))) // :: Stream Error String
  .map(S.lines)                                 // :: Stream Error (Array String)
  .map(S.map(path))                             // :: Stream Error (Array String)
  .map(S.map(readFileStream))                   // :: Stream Error (Array (Stream Error String))
  .chain(most.mergeArray)                       // :: String Error String
  .reduce(S.concat, '')                         // :: Promise Error String
```

Comment (Avaq) [inline most.js]: Let's keep the solution as is, I'm no frequent `most.js` user and cannot easily solve the "Promise settles before Most attaches handlers" problem. It's essentially just another Promise flaw - running into these issues really makes me appreciate Fluture. ;)

**Comment by davidchambers** [comment]
I'm about to squash the commits. The original commits can be found at [`e8e5d4f`][1].


[1]: https://github.com/plaid/async-problem/commits/e8e5d4f5f60193b1c1daa8dec49d8b8d18863a3b
